File: pkg/reposervice/download.go
package reposervice

import (
	"archive/zip"
	"context"
	"database/sql"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"connectrpc.com/connect"
	dbsqlc "github.com/ImSingee/git-plus/db/sqlc"
	"github.com/ImSingee/git-plus/pkg/repodownload"
	repov1 "github.com/ImSingee/git-plus/pkg/rpc/gitplus/repo/v1"
)

const repoDownloadRoutePrefix = "/repos/"

type downloadEstimate struct {
	archiveSizeBytes      int64
	estimatedProcessBytes int64
	estimatedDownloadSize int64
	processingLabel       string
}

type preparedDownload struct {
	rootPath string
	zipPath  string
	filename string
}

type stagedError struct {
	stage repov1.DownloadStage
	err   error
}

func (err stagedError) Error() string {
	return err.err.Error()
}

func (err stagedError) Unwrap() error {
	return err.err
}

func (s *serviceServer) StreamRepositoryDownload(
	ctx context.Context,
	req *connect.Request[repov1.StreamRepositoryDownloadRequest],
	stream *connect.ServerStream[repov1.StreamRepositoryDownloadResponse],
) error {
	repoID := req.Msg.GetRepoId()
	if repoID <= 0 {
		return connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("repo_id is required"))
	}

	queries, cleanup, err := s.openQueries(ctx)
	if err != nil {
		return connect.NewError(connect.CodeInternal, err)
	}
	defer cleanup()

	repo, err := queries.GetRepoById(ctx, repoID)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return connect.NewError(connect.CodeNotFound, fmt.Errorf("repository not found"))
		}
		return connect.NewError(connect.CodeInternal, fmt.Errorf("get repository: %w", err))
	}

	rows, err := queries.ListRepoRefsCurrentByRepoID(ctx, repoID)
	if err != nil {
		return connect.NewError(connect.CodeInternal, fmt.Errorf("list repository refs: %w", err))
	}

	activeRefs := filterActiveDownloadRefs(rows)
	estimate := estimateDownload(repo.ArchiveRepoSizeBytes)

	sendEvent := func(event *repov1.StreamRepositoryDownloadResponse) error {
		event.RepoId = int64Ptr(repoID)
		event.EstimatedProcessingLabel = stringPtr(estimate.processingLabel)
		if estimate.estimatedProcessBytes > 0 {
			event.EstimatedProcessingBytes = int64Ptr(estimate.estimatedProcessBytes)
		}
		if estimate.estimatedDownloadSize > 0 {
			event.EstimatedDownloadBytes = int64Ptr(estimate.estimatedDownloadSize)
		}
		if estimate.archiveSizeBytes > 0 {
			event.ArchiveSizeBytes = int64Ptr(estimate.archiveSizeBytes)
		}
		return stream.Send(event)
	}

	fail := func(stage repov1.DownloadStage, summary string, cause error) error {
		event := &repov1.StreamRepositoryDownloadResponse{
			State:           repov1.DownloadState_DOWNLOAD_STATE_FAILED.Enum(),
			Stage:           stage.Enum(),
			Summary:         stringPtr(summary),
			ProgressPercent: int32Ptr(100),
		}
		if cause != nil {
			event.ErrorMessage = stringPtr(cause.Error())
		}
		if err := sendEvent(event); err != nil {
			return err
		}
		return nil
	}

	if len(activeRefs) == 0 {
		return fail(
			repov1.DownloadStage_DOWNLOAD_STAGE_UNSPECIFIED,
			"No active branches or tags are available for download.",
			nil,
		)
	}

	reportProgress := func(stage repov1.DownloadStage, summary string, percent int32) error {
		return sendEvent(&repov1.StreamRepositoryDownloadResponse{
			State:           repov1.DownloadState_DOWNLOAD_STATE_RUNNING.Enum(),
			Stage:           stage.Enum(),
			Summary:         stringPtr(summary),
			ProgressPercent: int32Ptr(percent),
		})
	}

	artifact, err := s.prepareRepositoryDownload(ctx, repo, activeRefs, reportProgress)
	if err != nil {
		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
			return nil
		}
		stage := repov1.DownloadStage_DOWNLOAD_STAGE_UNSPECIFIED
		var stageErr stagedError
		if errors.As(err, &stageErr) {
			stage = stageErr.stage
		}
		return fail(stage, "Failed to prepare download.", err)
	}

	session := s.downloads.Register(repoID, artifact.zipPath, artifact.filename, artifact.rootPath)
	return sendEvent(&repov1.StreamRepositoryDownloadResponse{
		State:            repov1.DownloadState_DOWNLOAD_STATE_READY.Enum(),
		Stage:            repov1.DownloadStage_DOWNLOAD_STAGE_READY.Enum(),
		Summary:          stringPtr("Download is ready."),
		ProgressPercent:  int32Ptr(100),
		DownloadId:       stringPtr(session.ID),
		DownloadFilename: stringPtr(session.Filename),
	})
}

func (s *serviceServer) handleRepositoryDownload(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	repoID, downloadID, ok := parseDownloadRoute(r.URL.Path)
	if !ok {
		http.NotFound(w, r)
		return
	}

	session, err := s.downloads.Get(repoID, downloadID)
	if err != nil {
		switch {
		case errors.Is(err, repodownload.ErrSessionNotReady):
			http.Error(w, "download is not ready", http.StatusConflict)
		case errors.Is(err, repodownload.ErrRepoMismatch),
			errors.Is(err, repodownload.ErrNotFound),
			errors.Is(err, repodownload.ErrSessionExpired):
			http.NotFound(w, r)
		default:
			http.Error(w, "internal server error", http.StatusInternalServerError)
		}
		return
	}

	file, err := os.Open(session.ZipPath)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			http.NotFound(w, r)
			return
		}
		http.Error(w, "internal server error", http.StatusInternalServerError)
		return
	}
	defer file.Close()

	stat, err := file.Stat()
	if err != nil {
		http.Error(w, "internal server error", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/zip")
	w.Header().Set("Content-Disposition", contentDispositionFilename(session.Filename))
	http.ServeContent(w, r, session.Filename, stat.ModTime(), file)
}

func (s *serviceServer) prepareRepositoryDownload(
	ctx context.Context,
	repo dbsqlc.Repo,
	activeRefs []dbsqlc.RepoRefsCurrent,
	reportProgress func(stage repov1.DownloadStage, summary string, percent int32) error,
) (_ preparedDownload, err error) {
	sourcePath := filepath.Join(s.dataDir, "repos", repo.SourceID, repo.RefID)
	if _, statErr := os.Stat(sourcePath); statErr != nil {
		return preparedDownload{}, fmt.Errorf("archive repository not found: %w", statErr)
	}

	rootPath, err := os.MkdirTemp("", "git-plus-repo-download-*")
	if err != nil {
		return preparedDownload{}, fmt.Errorf("create temporary directory: %w", err)
	}

	cleanupRoot := true
	defer func() {
		if cleanupRoot {
			_ = os.RemoveAll(rootPath)
		}
	}()

	barePath := filepath.Join(rootPath, "archive.git")
	workPath := filepath.Join(rootPath, sanitizeRepoPathName(repo.Name))
	filename := buildDownloadFilename(repo)
	zipPath := filepath.Join(rootPath, filename)

	if err := reportProgress(repov1.DownloadStage_DOWNLOAD_STAGE_COPY_BARE, "Copying bare repository to a temporary directory...", 10); err != nil {
		return preparedDownload{}, err
	}
	if err := copyDirectory(ctx, sourcePath, barePath); err != nil {
		return preparedDownload{}, stagedError{
			stage: repov1.DownloadStage_DOWNLOAD_STAGE_COPY_BARE,
			err:   fmt.Errorf("copy bare repository: %w", err),
		}
	}

	if err := reportProgress(repov1.DownloadStage_DOWNLOAD_STAGE_MATERIALIZE_REFS, "Restoring active branches and tags...", 30); err != nil {
		return preparedDownload{}, err
	}
	headBranch, err := materializeActiveRefs(ctx, barePath, repo, activeRefs)
	if err != nil {
		return preparedDownload{}, stagedError{
			stage: repov1.DownloadStage_DOWNLOAD_STAGE_MATERIALIZE_REFS,
			err:   err,
		}
	}

	if err := reportProgress(repov1.DownloadStage_DOWNLOAD_STAGE_MATERIALIZE_REFS, "Creating a normal repository snapshot...", 60); err != nil {
		return preparedDownload{}, err
	}
	if err := cloneWorkingRepository(ctx, barePath, workPath); err != nil {
		return preparedDownload{}, stagedError{
			stage: repov1.DownloadStage_DOWNLOAD_STAGE_MATERIALIZE_REFS,
			err:   err,
		}
	}
	if err := createLocalBranches(ctx, workPath, activeRefs, headBranch); err != nil {
		return preparedDownload{}, stagedError{
			stage: repov1.DownloadStage_DOWNLOAD_STAGE_MATERIALIZE_REFS,
			err:   err,
		}
	}

	if err := reportProgress(repov1.DownloadStage_DOWNLOAD_STAGE_PACKAGE_ZIP, "Packaging repository archive...", 80); err != nil {
		return preparedDownload{}, err
	}
	if err := zipDirectory(ctx, workPath, zipPath, sanitizeRepoPathName(repo.Name)); err != nil {
		return preparedDownload{}, stagedError{
			stage: repov1.DownloadStage_DOWNLOAD_STAGE_PACKAGE_ZIP,
			err:   err,
		}
	}

	cleanupRoot = false
	return preparedDownload{
		rootPath: rootPath,
		zipPath:  zipPath,
		filename: filename,
	}, nil
}

func parseDownloadRoute(path string) (int64, string, bool) {
	trimmed := strings.TrimPrefix(path, repoDownloadRoutePrefix)
	parts := strings.Split(strings.Trim(trimmed, "/"), "/")
	if len(parts) != 4 || parts[1] != "downloads" || parts[3] != "archive" {
		return 0, "", false
	}

	repoID, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil || repoID <= 0 {
		return 0, "", false
	}

	downloadID := strings.TrimSpace(parts[2])
	if downloadID == "" {
		return 0, "", false
	}

	return repoID, downloadID, true
}

func filterActiveDownloadRefs(rows []dbsqlc.RepoRefsCurrent) []dbsqlc.RepoRefsCurrent {
	active := make([]dbsqlc.RepoRefsCurrent, 0, len(rows))
	for _, row := range rows {
		if row.Status != "active" {
			continue
		}
		if row.RefKind != "head" && row.RefKind != "tag" {
			continue
		}
		active = append(active, row)
	}
	sort.Slice(active, func(i int, j int) bool {
		return active[i].RefName < active[j].RefName
	})
	return active
}

func estimateDownload(value sql.NullInt64) downloadEstimate {
	if !value.Valid || value.Int64 <= 0 {
		return downloadEstimate{}
	}

	archiveSize := value.Int64
	return downloadEstimate{
		archiveSizeBytes:      archiveSize,
		estimatedProcessBytes: archiveSize,
		estimatedDownloadSize: int64(float64(archiveSize) * 0.7),
		processingLabel:       processingTimeEstimateLabel(archiveSize),
	}
}

func processingTimeEstimateLabel(sizeBytes int64) string {
	const (
		miB = 1024 * 1024
		giB = 1024 * miB
	)

	switch {
	case sizeBytes <= 50*miB:
		return "< 10s"
	case sizeBytes <= 250*miB:
		return "10-30s"
	case sizeBytes <= giB:
		return "30-90s"
	default:
		return "1-3 min"
	}
}

func materializeActiveRefs(ctx context.Context, barePath string, repo dbsqlc.Repo, activeRefs []dbsqlc.RepoRefsCurrent) (string, error) {
	branchNames := make([]string, 0)
	branchHashes := make(map[string]string)

	for _, ref := range activeRefs {
		if err := ctx.Err(); err != nil {
			return "", err
		}
		if err := runGit(ctx, filepath.Dir(barePath), "--git-dir", barePath, "update-ref", ref.RefName, ref.CurrentHash); err != nil {
			return "", fmt.Errorf("restore ref %q: %w", ref.RefName, err)
		}
		if ref.RefKind == "head" {
			shortName := strings.TrimPrefix(ref.RefName, "refs/heads/")
			branchNames = append(branchNames, shortName)
			branchHashes[shortName] = ref.CurrentHash
		}
	}

	sort.Strings(branchNames)
	headBranch := chooseHeadBranch(repo.DefaultBranch, branchNames)
	if headBranch != "" {
		if err := runGit(ctx, filepath.Dir(barePath), "--git-dir", barePath, "symbolic-ref", "HEAD", "refs/heads/"+headBranch); err != nil {
			return "", fmt.Errorf("set HEAD to %q: %w", headBranch, err)
		}
	}

	return headBranch, nil
}

func chooseHeadBranch(defaultBranch sql.NullString, branchNames []string) string {
	if len(branchNames) == 0 {
		return ""
	}

	if defaultBranch.Valid {
		trimmed := strings.TrimSpace(defaultBranch.String)
		for _, branchName := range branchNames {
			if branchName == trimmed {
				return branchName
			}
		}
	}

	return branchNames[0]
}

func cloneWorkingRepository(ctx context.Context, barePath string, workPath string) error {
	if err := runGit(ctx, filepath.Dir(workPath), "clone", barePath, workPath); err != nil {
		return fmt.Errorf("clone working repository: %w", err)
	}
	return nil
}

func createLocalBranches(ctx context.Context, workPath string, activeRefs []dbsqlc.RepoRefsCurrent, headBranch string) error {
	for _, ref := range activeRefs {
		if ref.RefKind != "head" {
			continue
		}
		branchName := strings.TrimPrefix(ref.RefName, "refs/heads/")
		if branchName == headBranch {
			continue
		}
		if err := runGit(ctx, workPath, "branch", "--force", branchName, ref.CurrentHash); err != nil {
			return fmt.Errorf("create local branch %q: %w", branchName, err)
		}
	}
	return nil
}

func copyDirectory(ctx context.Context, src string, dst string) error {
	return filepath.WalkDir(src, func(path string, d fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if err := ctx.Err(); err != nil {
			return err
		}

		relPath, err := filepath.Rel(src, path)
		if err != nil {
			return err
		}
		targetPath := dst
		if relPath != "." {
			targetPath = filepath.Join(dst, relPath)
		}

		info, err := d.Info()
		if err != nil {
			return err
		}

		switch {
		case d.IsDir():
			return os.MkdirAll(targetPath, info.Mode())
		case info.Mode()&os.ModeSymlink != 0:
			linkTarget, err := os.Readlink(path)
			if err != nil {
				return err
			}
			return os.Symlink(linkTarget, targetPath)
		default:
			return copyFile(path, targetPath, info.Mode())
		}
	})
}

func copyFile(src string, dst string, mode fs.FileMode) error {
	sourceFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer sourceFile.Close()

	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}

	targetFile, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, mode)
	if err != nil {
		return err
	}
	defer targetFile.Close()

	if _, err := io.Copy(targetFile, sourceFile); err != nil {
		return err
	}

	return nil
}

func zipDirectory(ctx context.Context, srcDir string, zipPath string, rootName string) error {
	file, err := os.Create(zipPath)
	if err != nil {
		return fmt.Errorf("create zip file: %w", err)
	}
	closeFile := true
	defer func() {
		if closeFile {
			_ = file.Close()
		}
	}()

	zipWriter := zip.NewWriter(file)
	closeZipWriter := true
	defer func() {
		if closeZipWriter {
			_ = zipWriter.Close()
		}
	}()

	if err := filepath.WalkDir(srcDir, func(path string, d fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if err := ctx.Err(); err != nil {
			return err
		}
		if path == srcDir {
			return nil
		}

		relPath, err := filepath.Rel(srcDir, path)
		if err != nil {
			return err
		}

		info, err := d.Info()
		if err != nil {
			return err
		}

		archivePath := filepath.ToSlash(filepath.Join(rootName, relPath))
		header, err := zip.FileInfoHeader(info)
		if err != nil {
			return err
		}
		header.Name = archivePath
		if d.IsDir() {
			header.Name += "/"
			_, err = zipWriter.CreateHeader(header)
			return err
		}

		if info.Mode()&os.ModeSymlink != 0 {
			header.Method = zip.Store
			writer, err := zipWriter.CreateHeader(header)
			if err != nil {
				return err
			}
			linkTarget, err := os.Readlink(path)
			if err != nil {
				return err
			}
			_, err = writer.Write([]byte(linkTarget))
			return err
		}

		header.Method = zip.Deflate
		writer, err := zipWriter.CreateHeader(header)
		if err != nil {
			return err
		}

		sourceFile, err := os.Open(path)
		if err != nil {
			return err
		}

		_, err = io.Copy(writer, sourceFile)
		closeErr := sourceFile.Close()
		if err != nil {
			return err
		}
		return closeErr
	}); err != nil {
		return fmt.Errorf("write zip archive: %w", err)
	}

	if err := zipWriter.Close(); err != nil {
		return fmt.Errorf("close zip writer: %w", err)
	}
	closeZipWriter = false
	if err := file.Close(); err != nil {
		return fmt.Errorf("close zip file: %w", err)
	}
	closeFile = false

	return nil
}

func runGit(ctx context.Context, dir string, args ...string) error {
	command := exec.CommandContext(ctx, "git", args...)
	command.Dir = dir
	output, err := command.CombinedOutput()
	if err != nil {
		return fmt.Errorf("%s: %w", strings.TrimSpace(string(output)), err)
	}
	return nil
}

func sanitizeRepoPathName(name string) string {
	trimmed := strings.TrimSpace(name)
	if trimmed == "" {
		return "repository"
	}
	replacer := strings.NewReplacer("/", "-", "\\", "-", ":", "-", "\x00", "")
	return replacer.Replace(trimmed)
}

func buildDownloadFilename(repo dbsqlc.Repo) string {
	base := strings.TrimSpace(repo.FullName)
	if base == "" {
		base = strings.TrimSpace(repo.Name)
	}
	base = sanitizeRepoPathName(base)
	return base + "-snapshot.zip"
}

func contentDispositionFilename(filename string) string {
	safe := strings.ReplaceAll(filename, "\"", "")
	return fmt.Sprintf("attachment; filename=\"%s\"", safe)
}

File: backend/src/download.ts
import { createReadStream, createWriteStream } from 'node:fs';
import { cp, mkdir, mkdtemp, readdir, rm, stat } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { basename, join, relative } from 'node:path';
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import yazl from 'yazl';
import { Code, ConnectError } from '@connectrpc/connect';
import {
  DownloadStage,
  DownloadState,
} from '../../frontend/src/rpc/gitplus/repo/v1/repo_pb';
import type { StreamRepositoryDownloadResponse } from '../../frontend/src/rpc/gitplus/repo/v1/repo_pb';
import type { FastifyReply, FastifyRequest } from 'fastify';
import type { AppDatabase, RepoRefRow, RepoRow } from './types';

const execFileAsync = promisify(execFile);
const SESSION_TTL_MS = 15 * 60 * 1000;

type DownloadSession = {
  id: string;
  repoId: number;
  zipPath: string;
  filename: string;
  rootPath: string;
  expiresAt: number;
};

export class DownloadManager {
  #sessions = new Map<string, DownloadSession>();

  register(
    repoId: number,
    zipPath: string,
    filename: string,
    rootPath: string,
  ): DownloadSession {
    this.cleanupExpired();
    const session = {
      id: crypto.randomUUID(),
      repoId,
      zipPath,
      filename,
      rootPath,
      expiresAt: Date.now() + SESSION_TTL_MS,
    };
    this.#sessions.set(session.id, session);
    return session;
  }

  get(repoId: number, id: string): DownloadSession {
    this.cleanupExpired();
    const session = this.#sessions.get(id);
    if (!session || session.repoId !== repoId) {
      throw new ConnectError('download session not found', Code.NotFound);
    }
    return session;
  }

  cleanupExpired(): void {
    const now = Date.now();
    for (const [id, session] of this.#sessions) {
      if (session.expiresAt <= now) {
        this.#sessions.delete(id);
        void rm(session.rootPath, { recursive: true, force: true });
      }
    }
  }
}

export async function* streamRepositoryDownload(
  dataDir: string,
  db: AppDatabase,
  downloads: DownloadManager,
  repoId: bigint,
): AsyncGenerator<StreamRepositoryDownloadResponse> {
  const id = Number(repoId);
  if (!Number.isSafeInteger(id) || id <= 0) {
    throw new ConnectError('repo_id is required', Code.InvalidArgument);
  }
  const repo = db.prepare('SELECT * FROM repos WHERE id = ?').get(id) as
    | RepoRow
    | undefined;
  if (!repo) throw new ConnectError('repository not found', Code.NotFound);

  const refs = db
    .prepare(
      "SELECT * FROM repo_refs_current WHERE repo_id = ? AND status = 'active' ORDER BY ref_name",
    )
    .all(id) as Array<RepoRefRow>;
  if (refs.length === 0) {
    yield event(
      id,
      DownloadState.FAILED,
      DownloadStage.UNSPECIFIED,
      'No active branches or tags are available for download.',
      100,
    );
    return;
  }

  try {
    yield event(
      id,
      DownloadState.RUNNING,
      DownloadStage.COPY_BARE,
      'Copying bare repository to a temporary directory...',
      10,
    );
    const artifact = await prepareRepositoryDownload(
      dataDir,
      repo,
      refs,
      (stage, summary, percent) => {
        void stage;
        void summary;
        void percent;
        return Promise.resolve();
      },
    );
    const session = downloads.register(
      id,
      artifact.zipPath,
      artifact.filename,
      artifact.rootPath,
    );
    yield {
      repoId,
      state: DownloadState.READY,
      stage: DownloadStage.READY,
      summary: 'Download is ready.',
      progressPercent: 100,
      downloadId: session.id,
      downloadFilename: session.filename,
    } as StreamRepositoryDownloadResponse;
  } catch (error) {
    yield {
      repoId,
      state: DownloadState.FAILED,
      stage: DownloadStage.UNSPECIFIED,
      summary: 'Failed to prepare download.',
      progressPercent: 100,
      errorMessage: error instanceof Error ? error.message : String(error),
    } as StreamRepositoryDownloadResponse;
  }
}

export async function handleDownloadRequest(
  downloads: DownloadManager,
  request: FastifyRequest,
  reply: FastifyReply,
): Promise<void> {
  const params = request.params as { repoId?: string; downloadId?: string };
  const repoId = Number(params.repoId);
  if (!Number.isSafeInteger(repoId) || !params.downloadId) {
    reply.code(404).send('not found');
    return;
  }
  const session = downloads.get(repoId, params.downloadId);
  reply.header('Content-Type', 'application/zip');
  reply.header(
    'Content-Disposition',
    `attachment; filename="${session.filename.replaceAll('"', '')}"`,
  );
  await reply.send(createReadStream(session.zipPath));
}

async function prepareRepositoryDownload(
  dataDir: string,
  repo: RepoRow,
  refs: Array<RepoRefRow>,
  progress: (
    stage: DownloadStage,
    summary: string,
    percent: number,
  ) => Promise<void>,
) {
  const sourcePath = join(dataDir, 'repos', repo.source_id, repo.ref_id);
  const rootPath = await mkdtemp(join(tmpdir(), 'git-plus-repo-download-'));
  const barePath = join(rootPath, 'archive.git');
  const workPath = join(rootPath, sanitize(repo.name));
  const filename = `${sanitize(repo.name)}-${new Date().toISOString().slice(0, 10)}.zip`;
  const zipPath = join(rootPath, filename);

  await progress(
    DownloadStage.COPY_BARE,
    'Copying bare repository to a temporary directory...',
    10,
  );
  await cp(sourcePath, barePath, { recursive: true });

  await progress(
    DownloadStage.MATERIALIZE_REFS,
    'Restoring active branches and tags...',
    30,
  );
  const headBranch = await materializeRefs(barePath, repo, refs);

  await progress(
    DownloadStage.MATERIALIZE_REFS,
    'Creating a normal repository snapshot...',
    60,
  );
  await runGit(rootPath, ['clone', barePath, workPath]);
  await createLocalBranches(workPath, refs, headBranch);

  await progress(
    DownloadStage.PACKAGE_ZIP,
    'Packaging repository archive...',
    80,
  );
  await zipDirectory(workPath, zipPath, sanitize(repo.name));
  return { rootPath, zipPath, filename };
}

async function materializeRefs(
  barePath: string,
  repo: RepoRow,
  refs: Array<RepoRefRow>,
): Promise<string> {
  const branches: Array<string> = [];
  for (const ref of refs) {
    await runGit(process.cwd(), [
      '--git-dir',
      barePath,
      'update-ref',
      ref.ref_name,
      ref.current_hash,
    ]);
    if (ref.ref_kind === 'head')
      branches.push(ref.ref_name.slice('refs/heads/'.length));
  }
  branches.sort();
  const headBranch = chooseHeadBranch(repo.default_branch, branches);
  if (headBranch) {
    await runGit(process.cwd(), [
      '--git-dir',
      barePath,
      'symbolic-ref',
      'HEAD',
      `refs/heads/${headBranch}`,
    ]);
  }
  return headBranch;
}

async function createLocalBranches(
  workPath: string,
  refs: Array<RepoRefRow>,
  headBranch: string,
): Promise<void> {
  for (const ref of refs) {
    if (ref.ref_kind !== 'head') continue;
    const branchName = ref.ref_name.slice('refs/heads/'.length);
    if (branchName === headBranch) continue;
    await runGit(workPath, ['branch', '--force', branchName, ref.current_hash]);
  }
}

function chooseHeadBranch(
  defaultBranch: string | null,
  branches: Array<string>,
): string {
  if (branches.length === 0) return '';
  if (defaultBranch && branches.includes(defaultBranch)) return defaultBranch;
  return branches[0] ?? '';
}

async function zipDirectory(
  srcDir: string,
  zipPath: string,
  rootName: string,
): Promise<void> {
  await mkdir(join(zipPath, '..'), { recursive: true });
  const zip = new yazl.ZipFile();
  const out = createWriteStream(zipPath);
  const done = new Promise<void>((resolve, reject) => {
    out.on('close', resolve);
    out.on('error', reject);
    zip.outputStream.on('error', reject);
  });
  zip.outputStream.pipe(out);
  await addDirectoryToZip(zip, srcDir, rootName);
  zip.end();
  await done;
}

async function addDirectoryToZip(
  zip: yazl.ZipFile,
  dir: string,
  rootName: string,
): Promise<void> {
  for (const entry of await readdir(dir, { withFileTypes: true })) {
    const path = join(dir, entry.name);
    const rel = join(rootName, relative(dir, path));
    if (entry.isDirectory()) {
      await addDirectoryToZip(zip, path, join(rootName, entry.name));
    } else if (entry.isFile()) {
      const fileStat = await stat(path);
      zip.addFile(path, rel, { mtime: fileStat.mtime, mode: fileStat.mode });
    }
  }
}

function event(
  repoId: number,
  state: DownloadState,
  stage: DownloadStage,
  summary: string,
  progressPercent: number,
): StreamRepositoryDownloadResponse {
  return {
    repoId: BigInt(repoId),
    state,
    stage,
    summary,
    progressPercent,
  } as StreamRepositoryDownloadResponse;
}

async function runGit(cwd: string, args: Array<string>): Promise<void> {
  await execFileAsync('git', args, { cwd, maxBuffer: 20 * 1024 * 1024 });
}

function sanitize(value: string): string {
  return (
    basename(value)
      .replace(/[^a-zA-Z0-9._-]+/g, '-')
      .replace(/^-+|-+$/g, '') || 'repo'
  );
}
